Close #17091

Dharshannan opened the pull request:
Close #17091

Dharshannan commented:
This PR is not complete, will need to add test function for ifftn frontend function.

Dharshannan commented:
3rd commit:
Fixed alignment issue and included statement to check if s shape and array shape are the same. It is difficult to pad and crop on the frontend as it results in inconsistent results between ivy.ifftn and numpy.fft.ifftn, RECOMMENDED : Code an ivy.ifftn backend function using numpy to allow for cropping and padding when the array shape and s argument shape do not match.

Dharshannan commented:
Commit 4:
Simplified and improved ifftn frontend function, now it is capable of handling zero padding and cropping via ivy.ifft.

Dharshannan commented:
Commit 5:
Removed ortho normalization handling in function as this is already done by ivy.ifft.

Dharshannan commented:
@zhumakhan Could I get a review for my current work please? Thanks.

Dharshannan commented:
Commit 6:
Added Test function for ifftn frontend
The ifftn frontend test has a 66% pass rate and fails for tensorflow backend, this might be caused by the fact that the ifft function also fails with tensorflow backend. ifft backend might have a problem and requires checking, since ifftn is coded using ifft, the failure with tensorflow backend is expected. Future Work: implement wider testing parameters (however this requires ifft fix for tensorflow backend to for better implementation).

ivy-leaves commented:
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

ivy-leaves commented:
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

Dharshannan commented:
Commit 7:
New ifftn function which works for a general case of any array, s and axes arguments!!!

Dharshannan commented:
@zhumakhan Could I get a review for my work so far please?

MuhammadNizamani commented on ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py:
```diff
@@ -189,3 +189,24 @@ def test_numpy_rfftfreq(n, sample_rate, frontend, test_flags, fn_tree, on_device
         n=n,
         d=d,
     )
+
+@handle_frontend_test(
```
lint is failing because there should be 2  lines space between tests function but your test function and other function has 1 line difference so make it

Dharshannan replied:
Lint test still fails, do I need pre-commit installed, for some reason pre-commit does not allow me to push changes I get an error:
![Pre-Commit_Error](https://github.com/unifyai/ivy/assets/94248626/3e57507b-2c14-445f-a4ca-72149f463748)
So I had to uninstall pre-commit with its hooks,

Dharshannan commented:
Commit 8:
Simplified the previously complex ifftn function.

Dharshannan commented:
Commit 9:
Updated test function, now tests have a 100% pass rate. Might have to code a different function similar to x_and_ifft(draw) to allow for data generation more catered towards the ifftn function.

Dharshannan commented:
Commit 10:
Changed frontend ifftn function to include backend ivy.ifftn which was newly added.

zhumakhan commented:
Hi, thanks for your PR!
It seems to me that ivy.ifftn should handle wrong inputs on its own. So there is no need to do input checks in the frontend.
```
@with_unsupported_dtypes({"1.24.3 and below": ("float16",)}, "numpy")
@to_ivy_arrays_and_back
def ifftn(a, s=None, axes=None, norm=None):
    a = ivy.array(a, dtype=ivy.complex128)
    a = ivy.ifftn(a, s=s, axes=axes, norm=norm)
    return a
``` 
This also passes all the tests.


In addition to that, lets check if there is better ways of typecasting the input array. Currently it is done as below:
```
    a = ivy.array(a, dtype=ivy.complex128)
```

Dharshannan commented:
> Hi, thanks for your PR! It seems to me that ivy.ifftn should handle wrong arguments on its own. So there is no need to do argument checks in the frontend.
> 
> ```
> @with_unsupported_dtypes({"1.24.3 and below": ("float16",)}, "numpy")
> @to_ivy_arrays_and_back
> def ifftn(a, s=None, axes=None, norm=None):
>     a = ivy.array(a, dtype=ivy.complex128)
>     a = ivy.ifftn(a, s=s, axes=axes, norm=norm)
>     return a
> ```
> 
> Above implementation passes all the tests.
> 
> In addition to that, lets check if there are better ways of typecasting the input array. Currently it is done as below:
> 
> ```
>     a = ivy.array(a, dtype=ivy.complex128)
> ```

@zhumakhan , should'nt the type casting be similar to that of the ifft function which is also ivy.complex128?

Dharshannan commented:
Commit 11:
Removed argument checks and changed type casting from "ivy.array" to "ivy.asarray" to handle type casting more flexibly and efficiently.

zhumakhan commented:
Hi, currently it is failing all of the tests. Could please update this PR accordingly to pass of the tests?
Thanks!

zhumakhan commented:
LGTM!
